subsys/dfu/src/modem_dfu_ctx.c

```diff
@@ -0,0 +1,199 @@
+	if (rc < 0) {
+		LOG_ERR("Failed to delete backup, errno %d\n", errno);
+		return 0;
+	}
+
+	k_delayed_work_submit_to_queue(&modem_fw_delete_offload_work_q,
+				       &modem_fw_delete_dwork, K_SECONDS(1));
+
+	return 0;
+}
+
+static void delete_old_modem_fw_task(struct k_work *w)
+{
+	int rc;
+	u32_t off;
+	nrf_socklen_t len = sizeof(off);
+
+	/* Request DFU offset from modem to ensure that the old firmware update
+	 * has been deleted to make space for the new.
+	 */
+	rc = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			    &off, &len);
+	if (rc == DFU_ERASE_PENDING) {
+		k_delayed_work_submit_to_queue(&modem_fw_delete_offload_work_q,
+					       &modem_fw_delete_dwork,
+					       K_MSEC(500));
+	} else if (rc < 0) {
+		LOG_ERR("Unable to delte old firmware image deleted");
+	} else {
+		LOG_INF("Old firmware image deleted");
+		k_sem_give(&modem_fw_delete_sem);
+	}
+}
+
+/**@brief Initialize DFU socket. */
+static int modem_dfu_socket_init(void)
+{
+	int rc;
+	nrf_socklen_t len;
+	nrf_dfu_fw_version_t version;
+	nrf_dfu_resources_t resource;
+	nrf_dfu_fw_offset_t off;
+
+	/* Create a socket for firmware update */
+	fd = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_STREAM, NRF_PROTO_DFU);
+	if (fd < 0) {
+		LOG_ERR("Failed to open Modem DFU socket.");
+		return fd;
+	}
+
+	LOG_INF("Modem DFU Socket created");
+
+	len = sizeof(version);
+	rc = nrf_getsockopt(fd, NRF_SOL_DFU,
+			    NRF_SO_DFU_FW_VERSION, &version, &len);
+	if (rc < 0) {
+		LOG_ERR("Firmware version request failed, errno %d", errno);
+		return -1;
+	}
+
+	*((u8_t *)version + sizeof(version) - 1) = '\0';
+	LOG_DBG("Modem firmware version: %s", log_strdup(version));
+
+	len = sizeof(resource);
+	rc = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_RESOURCES, &resource,
+			    &len);
+	if (rc < 0) {
+		LOG_ERR("Resource request failed, errno %d", errno);
+		return -1;
+	}
+
+	LOG_DBG("Flash memory available: %u", resource);
+
+	len = sizeof(off);
+	rc = nrf_getsockopt(fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			    &off, &len);
+	if (rc < 0) {
+		LOG_ERR("Offset request failed, errno %d", errno);
+		return -1;
+	}
+
+	LOG_DBG("Offset retrieved: %u", off);
+
+	return rc;
+}
+
+int modem_dfu_init(void)
+{
+	int err;
+
+	err = modem_dfu_socket_init();
+	if (err < 0) {
+		return err;
+	}
+
+	k_work_q_start(&modem_fw_delete_offload_work_q,
+		       modem_fw_delete_offload_work_q_stack,
+		       K_THREAD_STACK_SIZEOF(
+			       modem_fw_delete_offload_work_q_stack),
+		       CONFIG_MODEM_DFU_OFFLOAD_WORKQUEUE_PRIORITY);
+	k_sem_init(&modem_fw_delete_sem, 0, 1);
+	k_delayed_work_init(&modem_fw_delete_dwork, delete_old_modem_fw_task);
+
+	return 0;
+}
+
+int modem_dfu_write(u8_t *buf, size_t len)
+{
+	int sent = send(fd, buf, len, 0);
+
+	if (sent < 0) {
+		int modem_error = get_modem_error();
+
+		LOG_ERR("Modem refused fragment, errno %d, dfu err %d", errno,
+			modem_error);
+		if (modem_error == DFU_AREA_NOT_BLANK) {
+			/* Request delete task: TODO: Make sure the connection
```

Comment:
How can you identify what type of update it is before you have received information on what it is?

Reply:
I see; perhaps the best way to handle this is to do some self-diagnostic after installing a new modem firmware, to determine whether it can be deleted. That way you'd always be ready to receive a new image without having to wait to delete the previous one.

Reply:
I think the lather would be the best solution.